src/nvim/lua/xdiff.c

```diff
@@ -61,20 +66,90 @@ static int hunk_locations_cb(long start_a, long count_a, long start_b, long coun
   if (count_b > 0) {
     start_b += 1;
   }
-
   lua_State *lstate = (lua_State *)cb_data;
-  lua_createtable(lstate, 0, 0);
+  if (linematch) {
+    const char **diff_begin = xmalloc(sizeof(char *) * 2);
+    int *diff_length = xmalloc(sizeof(int) * 2);
+    diff_begin[0] = sma.ptr;
+    diff_length[0] = (int)count_a;
+    diff_begin[1] = smb.ptr;
+    diff_length[1] = (int)count_b;
+    // get the pointer to char of the start of the diff to pass it to linematch algorithm
+    for (int i = 0; i < 2; i++) {
+      for (int j = 0; j < (i ? start_b : start_a) - 1; j++) {
+        while (*diff_begin[i] != '\n') {
+          diff_begin[i]++;
+        }
+        diff_begin[i]++;
+      }
+    }
+    int decisions_length = 0;
+    int *decisions = NULL;
```

Comment:
Hmm, if you rework `linematch_nbuffers` to be callback based like xdiff, you could avoid many allocations.

Not essential for this PR however.